testnet/stacks-node/src/burnchains/bitcoin_regtest_controller.rs

```diff
@@ -59,9 +63,119 @@ pub struct BitcoinRegtestController {
     chain_tip: Option<BurnchainTip>,
     use_coordinator: Option<CoordinatorChannels>,
     burnchain_config: Option<Burnchain>,
-    last_utxos: Vec<UTXO>,
-    last_tx_len: u64,
-    min_relay_fee: u64, // satoshis/byte
+    ongoing_block_commit: Option<OngoingBlockCommit>,
+}
+
+struct OngoingBlockCommit {
+    payload: LeaderBlockCommitOp,
+    utxos: UTXOSet,
+    fees: LeaderBlockCommitFees,
+    txids: Vec<Txid>,
+}
+
+impl OngoingBlockCommit {
+    fn sum_utxos(&self) -> u64 {
+        self.utxos.total_available()
+    }
+}
+
+#[derive(Clone)]
+struct LeaderBlockCommitFees {
+    sunset_fee: u64,
+    fee_rate: u64,
+    sortition_fee: u64,
+    outputs_len: u64,
+    default_tx_size: u64,
+    spent_in_attempts: u64,
+    is_rbf_enabled: bool,
+    final_size: u64,
+}
+
+impl LeaderBlockCommitFees {
+    pub fn fees_from_previous_tx(
+        &self,
+        payload: &LeaderBlockCommitOp,
+        config: &Config,
+    ) -> LeaderBlockCommitFees {
+        let mut fees = LeaderBlockCommitFees::estimated_fees_from_payload(payload, config);
+        fees.spent_in_attempts = cmp::max(1, self.spent_in_attempts);
+        fees.final_size = self.final_size;
+        fees.fee_rate = self.fee_rate + config.burnchain.rbf_fee_increment;
+        fees.is_rbf_enabled = true;
+        fees
+    }
+
+    pub fn estimated_fees_from_payload(
```

Comment:
So, back in the world of Stacks 1.0 and `blockstack.js`, we spent the better part of a year trying to make fee estimation code like this work.  It ultimately failed -- no matter how refined our estimation heuristics were, they were never good enough to avoid corner cases where transactions didn't have enough fees to be mined.  We eventually scrapped them all and rewrite the code to calculate the transaction fee by _actually generating and signing_ the full transaction to get the payload length, and using _that_ to get the fee.

By far and away the biggest reasons for _not_ having an estimator like this are that you (1) cannot predict what kinds of UTXOs the transaction is going to consume, and (2) cannot predict how many UTXOs you need to consume to cover your fee.  If you get these wrong, then your transaction may not get mined in the next block, or even relayed at all.  The _only way_ you can reliably estimate your transaction fee is to try and assemble it from enough UTXOs that you can meet the targeted fee rate.

So, I think you should rewrite this method and estimate transaction fees only by actually generating them.

Reply:
yeah this is exactly the approach taken in this PR - see line 1111.

Reply:
The code uses this method as a starting estimation (line 100 and 892) before it tries to estimate the fee from the serialized transaction body.  This is sub-optimal -- the code should calculate the required fee only from a "real" transaction (for the reasons listed above).  It's a little tricker in this case, because the code also needs to budget for a maximum RBF fee bump, so it needs to pick the transaction's UTXOs accordingly before it does any estimations.